### DISCUSSION TITLE
Running optimizers without bindings

### DISCUSSION THREAD

**Discussion by dzfrias** [WebAssembly/binaryen#5870] [Q&A] [answered]
[`wasm-opt-rs`](https://github.com/brson/wasm-opt-rs), a bindings provider for Rust, doesn't use the C API to facilitate interactions between binaryen and Rust. It uses a C++ bindings generator, instead.

Right now, `wasm-opt-rs`'s output is not the same as binaryen's, when running at any optimization levels. In the process of fixing this behavior, I've tried to produce a minimal version of the [`wasm-opt.cpp`](https://github.com/WebAssembly/binaryen/blob/c003a01aa855bfa1377237eb2ce788b9fa96e839/src/tools/wasm-opt.cpp) file:

```c++
void optimize(std::string& infile, std::string& outfile) {
  // binaryen exports are namespaced under "wasm"
  wasm::Module wasm;
  wasm.features.enable(wasm::FeatureSet::Default);
  wasm.features.disable(wasm::FeatureSet::None);

  wasm::ModuleReader reader;
  reader.setDWARF(false);
  reader.setProfile(wasm::IRProfile::Normal);
  std::string inputSourceMapFilename;
  reader.read(infile, wasm, inputSourceMapFilename);

  auto passRunner = wasm::PassRunner(&wasm, wasm::PassOptions::getWithDefaultOptimizationOptions());
  passRunner.options.shrinkLevel = 2;
  passRunner.options.optimizeLevel = 2;
  passRunner.options.debugInfo = false;
  passRunner.addDefaultOptimizationPasses();
  passRunner.run();

  wasm::ModuleWriter writer;
  writer.setBinary(true);
  writer.setDebugInfo(false);
  writer.write(wasm, outfile);
}
```

Even with this, the outputs are not the same. The binary does definitely shrink, but not as drastically as it does with `wasm-opt`. I've noticed that even invoking `wasm-opt` without any optimization levels will lower the binary size by a sizeable amount. This doesn't seem to be the case with my minimal reproduction.

What part of the API is missing here? After looking through the codebase for quite a while I feel like I could be missing a small part of the API that enables the default behavior of `wasm-opt`.

**Comment by kripken** [comment]
I don't see anything obviously missing. Though you might want to experiment with changing the opt level to 3, and the shrink level to 1 or 0 (as 2 or 1 disable much inlining, which can have a large impact).

Otherwise, they way I would debug this is to run that code and the commandline tool with the same intended optimizations and with `BINARYEN_PASS_DEBUG=3` in the env, which will write out in the current directory a wasm file for each pass that is run. Using that output you can verify that the same passes were executed, and also see the changes in each one. The first file logged out (something like `byn-{PROCESS_ID}-000-before.wasm`) will be identical to the input, and the last one (`byn-{PROCESS_ID}-{LARGE_NUMBER}-..`) will be the final output file.

**Reply by dzfrias** [reply]
Thanks for the response! I tried running my test function in `tools/wasm-opt.cpp` after cloning this repository, and it worked as I intended...

Given that it works perfectly fine when not building it from `wasm-opt-rs`, I don't think it's related to the optimization passes (and they also seem to be running the same passes after setting `BINARYEN_PASS_DEBUG`). Just a clean run of `wasm-opt` (no `-O` arguments) shaves off around 50kb on my test input file that's 410kb large, which is really what I'm trying to reproduce. The only difference I can think of between the two repositories is the build system. As far as I know, that's the most probable reason why my test function works in this repository, but not in `wasm-opt-rs`.

The `wasm-opt-rs` bindings generator works by pulling in various sources from this repository and compiling them **without** `cmake`. Is there anything that the `cmake` build system for the repository does that influences the behavior of the compiled `wasm-opt` binary? I'll continue to look at `binaryen`'s build system to get a better idea of what it's doing.

Thanks again for your help, I appreciate the support.

**Reply by kripken** [reply]
I can't think of how CMake would influence how well optimizations work, that is very odd. Only theory I can have is that one is using older code that has fewer improvements?

But the `byn-*` files between two runs with different results should allow us to bisect this, since it shows step by step what the optimizer does. The first files must be identical between the runs, so any difference in the last files must start somewhere in the middle, which will point to a specific pass that behaves differently.

**Reply by dzfrias** [reply]
`wasm-opt-rs` uses `binaryen` version 114, commit 1fb1a2e2970472e9e93f9de94c8a2c674d0a0581. With an input of 410kb, `binaryen`'s version of `byn-000-before.wasm` has a size of 365kb. A ton of bytes seem to be shaved off even before the passes are run, which is strange. For `wasm-opt-rs`, the size of `byn-000-before.wasm` is 6 bytes over the size of the input.

**Reply by kripken** [reply]
Very strange. Any chance some other tool interposes in the middle?

If not, I'd diff those two files. Could be something obvious like one has a section the other doesn't, or different functions - whatever the diff is, it could be a big clue.

If you get stuck and can't figure out the 410kb => 365kb mystery using stock `wasm-opt`, I can take a look if you provide the binary.

**Reply by dzfrias** [reply]
Taking a look at the sections, the files are the exact same until the custom ones. There's one section that is much, much larger in the `wasm-opt-rs` version, and all the rest are the same. After stripping those sections, the files are identical. I'm still not quite sure why this is happening, but at least the problem's location is identified. Does `wasm-opt` do anything with custom sections that `wasm-opt-rs` wouldn't do?

On a slight side note, is it completely fine to strip custom sections in most production use-cases? As far as I know, WASM runtimes just ignore them. So unless a programmer is intending to use debug info from them, is it safe to assume that they can be stripped?

Also, in case you want it, here's the input file I'm using:

[input.wasm.zip](https://github.com/WebAssembly/binaryen/files/12317826/input.wasm.zip)

**Reply by dzfrias** [reply]
Currently, I'm trying to look at the bloated custom section in question. It's related to DWARF debug info. `wasm2wat` doesn't generate info about custom sections, I believe, so I'm not sure how to look into this part.

**Reply by dzfrias** [reply]
I was able to dump the DWARF-related custom sections. The dumped `.debug_line` section of the `wasm-opt-rs` version has two times the number of lines compared to the `wasm-opt` version. I found a [PR](https://github.com/WebAssembly/binaryen/pull/2545) relating to modifying DWARF-related custom sections. As far as I know, this works independently of the passes(?), which might explain the sudden 410kb to 365kb jump.

After diffing the dumps, I found all the DWARF line tables are completely empty in the `.debug_line` section in the `wasm-opt` version. They only display the line table metadata, and don't have any actual mappings. Is this the intended behavior?

Aside from the empty line tables, I took a look at [`wasm-debug.cpp`](https://github.com/WebAssembly/binaryen/blob/616c08bc1ec604a822d354cbb0353b7994cec72d/src/wasm/wasm-debug.cpp) and didn't get any indication as to why the behavior changes in `wasm-opt-rs`. There's a big `#ifdef BUILD_LLVM_DWARF` surrounding most of the code, but switching this option off by default in the `CMakeLists.txt` file didn't change much.

edit:
I went back and tried to not build with `-DBUILD_LLVM_DWARF` again, and it turns out I didn't actually turn it off in `CMakeLists.txt` the first time around. After doing that, the binary sizes of `wasm-opt` and `wasm-opt-rs` finally matched. I believe this is a bug in `wasm-opt-rs`, as full conformity was one of the original intents of the project. Although the root original issue presented in this discussion has been found, I'm still curious about the empty DWARF line tables produced by `wasm-opt`. Again, I'm not sure if this is intended or not.

Regardless, thank you for the continuous help and guidance in this discussion!

**Reply by kripken** [reply, marked as answer]
Ah, if it is a DWARF section that CMake does in fact affect that. That's not an optimization issue but a matter of debug support. Makes sense now.

About the line tables, Binaryen's DWARF support is somewhat limited so it's possible you're hitting a limitation or bug. In general it should work when few optimizations are run, however.

**Reply by dzfrias** [reply]
Good to know, thank you. I'll close up this discussion now!
